flink-table/flink-table-code-splitter/src/main/java/org/apache/flink/table/codesplit/BlockStatementGrouper.java

```diff
@@ -0,0 +1,432 @@
+                rewriter.replace(
+                        value.get(0).getStart(),
+                        value.get(value.size() - 1).getStop(),
+                        group.getKey() + "(" + this.parameters + ");");
+            }
+
+            return groups;
+        }
+    }
+
+    private interface LocalGroupElement {
+
+        Token getStart();
+
+        Token getStop();
+
+        String getBody();
+
+        ParserRuleContext getContext();
+    }
+
+    private static class ContextGroupElement implements LocalGroupElement {
+
+        private final ParserRuleContext parserRuleContext;
+
+        private ContextGroupElement(ParserRuleContext parserRuleContext) {
+            this.parserRuleContext = parserRuleContext;
+        }
+
+        @Override
+        public Token getStart() {
+            return this.parserRuleContext.start;
+        }
+
+        @Override
+        public Token getStop() {
+            return this.parserRuleContext.stop;
+        }
+
+        @Override
+        public String getBody() {
+            return CodeSplitUtil.getContextString(this.parserRuleContext);
+        }
+
+        @Override
+        public ParserRuleContext getContext() {
+            return this.parserRuleContext;
+        }
+    }
+
+    private static class RewriteContextGroupElement implements LocalGroupElement {
+
+        private final ParserRuleContext parserRuleContext;
+
+        private final TokenStreamRewriter rewriter;
+
+        private RewriteContextGroupElement(
+                ParserRuleContext parserRuleContext, TokenStreamRewriter rewriter) {
+            this.parserRuleContext = parserRuleContext;
+            this.rewriter = rewriter;
+        }
+
+        @Override
+        public Token getStart() {
+            return this.parserRuleContext.start;
+        }
+
+        @Override
+        public Token getStop() {
+            return this.parserRuleContext.stop;
+        }
+
+        @Override
+        public String getBody() {
+            return this.rewriter.getText();
+        }
+
+        @Override
+        public ParserRuleContext getContext() {
+            return this.parserRuleContext;
+        }
+    }
+
+    private static class ReturnCounter extends JavaParserBaseVisitor<Void> {
+
+        private int returnCount = 0;
+
+        @Override
+        public Void visitStatement(StatementContext ctx) {
+            if (ctx.RETURN() != null) {
+                returnCount++;
+            }
+            return visitChildren(ctx);
+        }
+    }
+
+    /**
+     * This object represents a rewritten code block. It contains its new form along with all
+     * extracted groups and their names.
+     */
+    public static class RewriteGroupedCode {
+
+        /** Rewritten code block containing calls to extracted methods. */
+        private final String rewriteCode;
+
+        /** All extracted groups with their names. */
+        private final List<Pair<String, List<String>>> groups;
```

Comment:
Hi,
I always thought about Flink's  Tuple classes as designed to use in Flink's DataStream Applications rather than core/internal implementations. 

I did a quick search in Flink's base code and I see that `org.apache.commons.lang3.tuple.Pair` is used in quite few places like:
Hive-connector
CoGroupOperator (flink-java)
Few classes from Java-runtime module
Flink-table module

That's why I considered it as well suited class since it is already on a class path.